xds/internal/xdsclient/authority_test.go

```diff
@@ -0,0 +1,249 @@
+/*
+ *
+ * Copyright 2023 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package xdsclient
+
+import (
+	"context"
+	"testing"
+
+	v3corepb "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	v3listenerpb "github.com/envoyproxy/go-control-plane/envoy/config/listener/v3"
+	"github.com/google/uuid"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/internal/testutils/xds/e2e"
+	"google.golang.org/grpc/xds/internal"
+	"google.golang.org/grpc/xds/internal/httpfilter/router"
+	"google.golang.org/grpc/xds/internal/xdsclient/bootstrap"
+	"google.golang.org/grpc/xds/internal/xdsclient/xdsresource"
+	"google.golang.org/grpc/xds/internal/xdsclient/xdsresource/version"
+)
+
+type delegatingDummyWatcher struct {
+	watcher listenerWatcher
+}
+
+func (d *delegatingDummyWatcher) OnUpdate(data xdsresource.ResourceData) {
+	l := data.(*xdsresource.ListenerResourceData)
+	d.watcher.OnUpdate(l)
+}
+
+func (d *delegatingDummyWatcher) OnError(err error) {
+	d.watcher.OnError(err)
+}
+
+func (d *delegatingDummyWatcher) OnResourceDoesNotExist() {
+	d.watcher.OnResourceDoesNotExist()
+}
+
+var (
+	rType      = internal.ResourceTypeMapForTesting[version.V3ListenerURL].(xdsresource.Type)
+	rtRegistry = newResourceTypeRegistry()
+	_          = router.TypeURL
+)
+
+func init() {
+	// Simulating maybeRegister for rType.
+	rtRegistry.types[rType.TypeURL()] = rType
+}
```

Comment:
This diff is not part of this PR. This would be added after this [PR](https://github.com/grpc/grpc-go/pull/5996) is merged in.